File: app/src/main/java/com/miearn/app/domain/LearningSession.kt
package com.miearn.app.domain

enum class LearningPhase {
    REVIEW,
    BROWSE,
    CONSOLIDATE,
    REINFORCEMENT,
    COMPLETE,
}

data class LearningSession(
    val reviewIds: List<String>,
    val newIds: List<String>,
    val reinforcementIds: List<String>,
    val phase: LearningPhase,
    val index: Int,
    val completedNew: Int,
    val completedReview: Int,
    val correctFirstTry: Int,
    val answeredFirstTry: Int,
    val cardExpanded: Boolean,
    val pendingFirstCorrect: Boolean?,
) {
    val currentId: String?
        get() = when (phase) {
            LearningPhase.REVIEW -> reviewIds.getOrNull(index)
            LearningPhase.BROWSE, LearningPhase.CONSOLIDATE -> newIds.getOrNull(index)
            LearningPhase.REINFORCEMENT -> reinforcementIds.getOrNull(index)
            LearningPhase.COMPLETE -> null
        }

    val isComplete: Boolean
        get() = phase == LearningPhase.COMPLETE

    val phaseTotal: Int
        get() = when (phase) {
            LearningPhase.REVIEW -> reviewIds.size
            LearningPhase.BROWSE, LearningPhase.CONSOLIDATE -> newIds.size
            LearningPhase.REINFORCEMENT -> reinforcementIds.size
            LearningPhase.COMPLETE -> 0
        }

    fun submitAnswer(firstCorrect: Boolean): LearningSession {
        if (
            phase !in ANSWER_PHASES ||
            currentId == null ||
            pendingFirstCorrect != null
        ) {
            return this
        }
        val firstTry = phase != LearningPhase.REINFORCEMENT
        val nextReinforcement = if (!firstCorrect && currentId !in reinforcementIds) {
            reinforcementIds + checkNotNull(currentId)
        } else {
            reinforcementIds
        }
        return copy(
            reinforcementIds = nextReinforcement,
            completedNew = completedNew +
                if (firstTry && phase == LearningPhase.CONSOLIDATE) 1 else 0,
            completedReview = completedReview +
                if (firstTry && phase == LearningPhase.REVIEW) 1 else 0,
            correctFirstTry = correctFirstTry +
                if (firstTry && firstCorrect) 1 else 0,
            answeredFirstTry = answeredFirstTry + if (firstTry) 1 else 0,
            pendingFirstCorrect = firstCorrect,
        )
    }

    fun continueAfterAnswer(): LearningSession {
        if (phase !in ANSWER_PHASES || pendingFirstCorrect == null) return this
        return when (phase) {
            LearningPhase.REVIEW -> {
                if (index + 1 < reviewIds.size) {
                    copy(index = index + 1, pendingFirstCorrect = null)
                } else {
                    transitionAfterReview()
                }
            }

            LearningPhase.CONSOLIDATE -> {
                if (index + 1 < newIds.size) {
                    copy(index = index + 1, pendingFirstCorrect = null)
                } else {
                    transitionToReinforcementOrComplete()
                }
            }

            LearningPhase.REINFORCEMENT -> {
                if (index + 1 < reinforcementIds.size) {
                    copy(index = index + 1, pendingFirstCorrect = null)
                } else {
                    complete()
                }
            }

            else -> this
        }
    }

    fun nextBrowse(): LearningSession {
        if (phase != LearningPhase.BROWSE || newIds.isEmpty()) return this
        return if (index + 1 < newIds.size) {
            copy(index = index + 1, cardExpanded = false)
        } else {
            copy(
                phase = LearningPhase.CONSOLIDATE,
                index = 0,
                cardExpanded = false,
                pendingFirstCorrect = null,
            )
        }
    }

    fun previousBrowse(): LearningSession =
        if (phase == LearningPhase.BROWSE && index > 0) {
            copy(index = index - 1, cardExpanded = false)
        } else {
            this
        }

    fun browseTo(targetIndex: Int): LearningSession =
        if (phase == LearningPhase.BROWSE && targetIndex in newIds.indices) {
            copy(index = targetIndex, cardExpanded = false)
        } else {
            this
        }

    fun toggleCardExpanded(): LearningSession =
        if (phase == LearningPhase.BROWSE) {
            copy(cardExpanded = !cardExpanded)
        } else {
            this
        }

    private fun transitionAfterReview(): LearningSession =
        if (newIds.isNotEmpty()) {
            copy(
                phase = LearningPhase.BROWSE,
                index = 0,
                cardExpanded = false,
                pendingFirstCorrect = null,
            )
        } else {
            transitionToReinforcementOrComplete()
        }

    private fun transitionToReinforcementOrComplete(): LearningSession =
        if (reinforcementIds.isNotEmpty()) {
            copy(
                phase = LearningPhase.REINFORCEMENT,
                index = 0,
                cardExpanded = false,
                pendingFirstCorrect = null,
            )
        } else {
            complete()
        }

    private fun complete(): LearningSession =
        copy(
            phase = LearningPhase.COMPLETE,
            index = 0,
            cardExpanded = false,
            pendingFirstCorrect = null,
        )

    companion object {
        private val ANSWER_PHASES = setOf(
            LearningPhase.REVIEW,
            LearningPhase.CONSOLIDATE,
            LearningPhase.REINFORCEMENT,
        )

        fun start(
            reviewIds: List<String>,
            newIds: List<String>,
        ): LearningSession {
            val phase = when {
                reviewIds.isNotEmpty() -> LearningPhase.REVIEW
                newIds.isNotEmpty() -> LearningPhase.BROWSE
                else -> LearningPhase.COMPLETE
            }
            return LearningSession(
                reviewIds = reviewIds,
                newIds = newIds,
                reinforcementIds = emptyList(),
                phase = phase,
                index = 0,
                completedNew = 0,
                completedReview = 0,
                correctFirstTry = 0,
                answeredFirstTry = 0,
                cardExpanded = false,
                pendingFirstCorrect = null,
            )
        }
    }
}

File: ios/MIearnCore/Sources/MIearnCore/LearningSession.swift
import Foundation

public enum LearningPhase: String, Codable, Sendable, Hashable {
    case review
    case browse
    case consolidate
    case reinforcement
    case complete
}

public struct LearningSession: Codable, Equatable, Sendable {
    public var reviewIds: [String]
    public var newIds: [String]
    public var reinforcementIds: [String]
    public var phase: LearningPhase
    public var index: Int
    public var completedNew: Int
    public var completedReview: Int
    public var correctFirstTry: Int
    public var answeredFirstTry: Int
    public var cardExpanded: Bool
    public var pendingFirstCorrect: Bool?

    public var currentId: String? {
        switch phase {
        case .review:
            return reviewIds[safe: index]
        case .browse, .consolidate:
            return newIds[safe: index]
        case .reinforcement:
            return reinforcementIds[safe: index]
        case .complete:
            return nil
        }
    }

    public var isComplete: Bool { phase == .complete }

    public var phaseTotal: Int {
        switch phase {
        case .review:
            return reviewIds.count
        case .browse, .consolidate:
            return newIds.count
        case .reinforcement:
            return reinforcementIds.count
        case .complete:
            return 0
        }
    }

    public static func start(reviewIds: [String], newIds: [String]) -> LearningSession {
        let phase: LearningPhase
        if !reviewIds.isEmpty {
            phase = .review
        } else if !newIds.isEmpty {
            phase = .browse
        } else {
            phase = .complete
        }
        return LearningSession(
            reviewIds: reviewIds,
            newIds: newIds,
            reinforcementIds: [],
            phase: phase,
            index: 0,
            completedNew: 0,
            completedReview: 0,
            correctFirstTry: 0,
            answeredFirstTry: 0,
            cardExpanded: false,
            pendingFirstCorrect: nil
        )
    }

    public func submitAnswer(firstCorrect: Bool) -> LearningSession {
        guard Self.answerPhases.contains(phase),
              let currentId,
              pendingFirstCorrect == nil else {
            return self
        }
        let firstTry = phase != .reinforcement
        var next = self
        if !firstCorrect && !next.reinforcementIds.contains(currentId) {
            next.reinforcementIds.append(currentId)
        }
        if firstTry && phase == .consolidate {
            next.completedNew += 1
        }
        if firstTry && phase == .review {
            next.completedReview += 1
        }
        if firstTry && firstCorrect {
            next.correctFirstTry += 1
        }
        if firstTry {
            next.answeredFirstTry += 1
        }
        next.pendingFirstCorrect = firstCorrect
        return next
    }

    public func continueAfterAnswer() -> LearningSession {
        guard Self.answerPhases.contains(phase), pendingFirstCorrect != nil else {
            return self
        }
        switch phase {
        case .review:
            if index + 1 < reviewIds.count {
                return updatingIndex(index + 1)
            }
            return transitionAfterReview()
        case .consolidate:
            if index + 1 < newIds.count {
                return updatingIndex(index + 1)
            }
            return transitionToReinforcementOrComplete()
        case .reinforcement:
            if index + 1 < reinforcementIds.count {
                return updatingIndex(index + 1)
            }
            return completing()
        case .browse, .complete:
            return self
        }
    }

    public func nextBrowse() -> LearningSession {
        guard phase == .browse, !newIds.isEmpty else { return self }
        var next = self
        if index + 1 < newIds.count {
            next.index += 1
        } else {
            next.phase = .consolidate
            next.index = 0
            next.pendingFirstCorrect = nil
        }
        next.cardExpanded = false
        return next
    }

    public func previousBrowse() -> LearningSession {
        guard phase == .browse, index > 0 else { return self }
        var next = self
        next.index -= 1
        next.cardExpanded = false
        return next
    }

    public func browseTo(_ targetIndex: Int) -> LearningSession {
        guard phase == .browse, newIds.indices.contains(targetIndex) else { return self }
        var next = self
        next.index = targetIndex
        next.cardExpanded = false
        return next
    }

    public func toggleCardExpanded() -> LearningSession {
        guard phase == .browse else { return self }
        var next = self
        next.cardExpanded.toggle()
        return next
    }

    private func updatingIndex(_ newIndex: Int) -> LearningSession {
        var next = self
        next.index = newIndex
        next.pendingFirstCorrect = nil
        return next
    }

    private func transitionAfterReview() -> LearningSession {
        if !newIds.isEmpty {
            var next = self
            next.phase = .browse
            next.index = 0
            next.cardExpanded = false
            next.pendingFirstCorrect = nil
            return next
        }
        return transitionToReinforcementOrComplete()
    }

    private func transitionToReinforcementOrComplete() -> LearningSession {
        guard !reinforcementIds.isEmpty else { return completing() }
        var next = self
        next.phase = .reinforcement
        next.index = 0
        next.cardExpanded = false
        next.pendingFirstCorrect = nil
        return next
    }

    private func completing() -> LearningSession {
        var next = self
        next.phase = .complete
        next.index = 0
        next.cardExpanded = false
        next.pendingFirstCorrect = nil
        return next
    }

    private static let answerPhases: Set<LearningPhase> = [
        .review,
        .consolidate,
        .reinforcement,
    ]
}

private extension Array {
    subscript(safe index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
